### ISSUE TITLE
Рекомендации по доработке драйвера XGBE

### ISSUE THREAD

**Issue by wmdlr** [wmdlr/linux-kernel#4]
Доработку предлагаю проводить поэтапно.

На начальном этапе (**эксперимент №1**) предлагаю:
1. Для приема и передачи пакетов использовать единственную структуру struct napi на канал. Для этого в структуре _struct xgbe_channel_ поля _napi_rx_ и _napi_tx_ заменить на _struct napi_struct napi_.
2. Для каждого канала настраивать единственный обработчик прерываний DMA. Для этого в структуре _struct xgbe_channel_ поля _tx_dma_irq_ и _rx_dma_irq_ заменить на _int dma_irq_. Обработчик прерываний DMA реализовать в виде:
```
static irqreturn_t xgbe_dma_isr(int irq, void *data) {
	struct xgbe_channel *channel = data;
	unsigned int dma_ch_isr;

	dma_ch_isr = XGMAC_DMA_IOREAD(channel, DMA_CH_SR);

	if (napi_schedule_prep(&channel->napi)) {
		disable_irq_nosync(channel->dma_irq);
		__napi_schedule(&channel->napi);
	}

	XGMAC_DMA_IOWRITE(channel, DMA_CH_SR, dma_ch_isr);

        return IRQ_HANDLED;
}
```
3. Вместо функций _xgbe_one_tx_poll()_ и _xgbe_one_rx_poll()_ реализовать функцию _xgbe_one_poll()_ следующего вида:
```
static int xgbe_one_poll(struct napi_struct *napi, int budget)
{
	struct xgbe_channel *channel = container_of(napi, struct xgbe_channel, napi);
	int processed = 0;

	xgbe_tx_poll(channel);

	processed = xgbe_rx_poll(channel, budget);

	if (processed < budget) {
		napi_complete(napi);
		enable_irq(channel->dma_irq);
	}

	return processed;
}
```
4. Уменьшить значение _XGBE_NAPI_POLL_WEIGHT_ до _NAPI_POLL_WEIGHT_. Адекватное значение установить экспериментально на основании измеренной частоты запусков функции _xgbe_one_poll()_ под нагрузкой при установленном нулевом бюджете.

**Comment by vmitrofanov** [comment]
Данные рекомендации очень похожи на конфигурацию драйвера с выключенным per_channel_irq. Ввиду ее схожести изменения буду вносить в нее.

**Comment by wmdlr** [comment]
> Данные рекомендации очень похожи на конфигурацию драйвера с выключенным per_channel_irq. Ввиду ее схожести изменения буду вносить в нее.

Да, похожи. Но мы все-таки оставляем раздельную обработку каналов.

**Comment by vmitrofanov** [comment]
> Да, похожи. Но мы все-таки оставляем раздельную обработку каналов.
Теперь понял тебя.

**Comment by vmitrofanov** [comment]
Отправил код для ознакомления, можно пока не пушить pull_request, потому, что не проверил эти изменения. Возникли проблемы с образом, видимо, что то не запушилось или запушилось не то, что надо на начальных коммитах... eth0 ведет себя странным образом. Пока проверяю, где что-сломалось

**Comment by wmdlr** [comment]
Попутно возник вопрос.
В функции _xgbe_request_irqs()_ идет настройка обработчиков прерываний по приему и передаче для каждого канала DMA. В каком случае они срабатывают и влияют ли на них настройки регистра **DMA_CHn_Interrupt_Enable**?
Поясню. В регистре **DMA_CHn_Interrupt_Enable** устанавливаются различные источники прерываний. Например, чисто для приема можно задать **RSE**, **RBUE**, **RIE**. Какие из них приводят к срабатыванию канального прерывания по приему? Аналогично и для передачи.

**Comment by vmitrofanov** [comment]
Используются только RIE и TIE

**Comment by vmitrofanov** [comment]
>В каком случае они срабатывают и влияют ли на них настройки регистра DMA_CHn_Interrupt_Enable?

TI (Bit0): This bit indicatis that the packet transmission is complete. When transmission is complete, Bit 31 of TDES1 is reset in the first descriptor, and the specific packet status information is updated in the descriptor.... Bit 0 of status register of corresponding channel is set after completing transmission of a packet that has interrupt on completion (TEDES2[31]) set in its last descriptor. the DMA engine return to step 3 if the descriptor tail pointer offset is not equal to the current descriptor closed by the DMA

RI: This bin indicates that the packet reception is complete. When packet reception is complete, Git 31 of RDES1 is reset in the last descriptor, and the specific packet status information is updated in the descriptor. The reception remains in the Running state

**Comment by wmdlr** [comment]
> Используются только RIE и TIE

Тогда требуются еще пояснения. В функции _xgbe_request_irqs()_ устанавливаются 5 обработчиков прерываний. Один (интегральный) для всего контроллера, остальные - по 2 на каждый канал. Вот распределение для платы Топаз:
```
 81:          3          0  MIPS GIC  81  eth0
 82:         29          0  MIPS GIC  82  eth0-Tx-0
 83:          0          0  MIPS GIC  83  eth0-Tx-1
 84:          0          0  MIPS GIC  84  eth0-Rx-0
 85:          0          0  MIPS GIC  85  eth0-Rx-1
```
Интегральное прерывание 81 будет (или может) срабатывать при приеме и передаче пакетов? Т.е. одновременно с каким-либо из множества 82-85? Или же в контроллере XGMAC существует четкое функциональное разделение прерываний, при котором прием пакетов попадает только на 84 и 85, передача - только на 82 и 83, а все остальное - на 81?

**Comment by vmitrofanov** [comment]
> Используются только RIE и TIE

Перерепроверил еще раз. Действительно это не так. Все что описано ниже касается режима per_channel_interrupt.
О разрешенных прерываниях написано ясно в xgbe-dev.c -> xgbe_eneable_dma_interrupts():
Там выставляются флаги: 
* NIE (Normal Interrupt Summary Enable),  он активирует TIE, TBU, RIE
* AIE (Abnormal Interrupt Summary Enable), он активирует TXSE, TBUE, RSE, RBUE, CDEE
* FBEE (Fatal Bus Error Enable), он активирует AIE

Устанавливается  поле INTM в 1 (отвечает за поведения сигнала прерывания): sbd_perch_tx, sbd_perch_rx are level signals asserted on corresponding event and de-asserted when the software clears the corresponding RI/TI status bits sbd_intr_o is not asserted for these packet transfer completion events. sbd_intr_o - the common output signal.

Таким образом "общий" xgbe_isr (81) должен возникать во всех случаях кроме TI и RI, получается только для обработки исключительных ситуаций.

На всякий случай повторюсь: TI устанавливается когда DMA канал закрывает дескриптор с установленным IOC (Interrupt on Completion -TDES2[31])
RI - когда установлен LD (Last Descriptor) и IOC.

**Comment by wmdlr** [comment]
> Очень похоже, что вылеты происходят при одновременном вызове xgbe_tx_poll() из xgbe_tx_timer, который вызывается в xgbe_xmit() и одновременном вызове xgbe_tx_poll() из xgbe_one_poll(), который происходит по прерыванию RX (происходит одновременный вызов unmap_rdata). Если добавить мьютекс для только отдельного доступа, то все работает ок.

Да, согласен, пропустил я момент с состоянием гонки. Не хочется, если честно, использовать средства синхронизации. Предлагаю не вводить их дополнительно в _xgbe_tx_poll()_, а _xgbe_tx_timer()_ переписать в виде:
```
static void xgbe_tx_timer(unsigned long data)
{
	struct xgbe_channel *channel = (struct xgbe_channel *)data;
	struct xgbe_prv_data *pdata = channel->pdata;
	struct napi_struct *napi;

	DBGPR("-->xgbe_tx_timer\n");

	napi = pdata->per_channel_irq ? &channel->napi : &pdata->napi;

	if (napi_schedule_prep(napi)) {
		if (!pdata->per_channel_irq) {
			xgbe_disable_rx_tx_ints(pdata);
		} else {
			disable_irq_nosync(channel->rx_dma_irq);
			disable_irq_nosync(channel->tx_dma_irq);
		}

		/* Turn on polling */
		__napi_schedule(napi);
	}

	channel->tx_timer_active = 0;

	DBGPR("<--xgbe_tx_timer\n");
}
```
Функцию _xgbe_dma_isr()_ тогда немного дополнить:
```
static irqreturn_t xgbe_dma_isr(int irq, void *data) {

	struct xgbe_channel *channel = data;
	unsigned int dma_ch_isr;

	dma_ch_isr = XGMAC_DMA_IOREAD(channel, DMA_CH_SR);

	if (napi_schedule_prep(&channel->napi)) {
		disable_irq_nosync(channel->rx_dma_irq);
		disable_irq_nosync(channel->tx_dma_irq);

		/* Turn on polling */
		__napi_schedule(&channel->napi);
	}

	XGMAC_DMA_IOWRITE(channel, DMA_CH_SR, dma_ch_isr);

	return IRQ_HANDLED;
}
```
Аналогично дополнить и _xgbe_one_poll()_:
```
static int xgbe_one_poll(struct napi_struct *napi, int budget)
{
	struct xgbe_channel *channel = container_of(napi, struct xgbe_channel, napi);
	int processed;

	xgbe_tx_poll(channel);

	processed = xgbe_rx_poll(channel, budget);

	if (processed < budget) {
		napi_complete(napi);

		enable_irq(channel->rx_dma_irq);
		enable_irq(channel->tx_dma_irq);
	}

	return processed;
}
```
И в _xgbe_all_poll()_ раскомментировать вызов _xgbe_enable_rx_tx_ints()_.

**Comment by vmitrofanov** [comment]
см. ветку vasily_recomendation. Лучше мержить ее
